sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,766 @@
+
+            X_binned_train, X_binned_val, y_train, y_val = train_test_split(
+                X_binned, y, test_size=self.validation_fraction,
+                stratify=stratify, random_state=rng)
+            if X_binned_train.size == 0 or X_binned_val.size == 0:
+                raise ValueError(
+                    'Not enough data (n_samples={}) to '
+                    'perform early stopping with validation_fraction='
+                    '{}. Use more training data or '
+                    'adjust validation_fraction.'.format(
+                        X_binned.shape[0],
+                        self.validation_fraction)
+                )
+            # Predicting is faster of C-contiguous arrays, training is faster
+            # on Fortran arrays.
+            X_binned_val = np.ascontiguousarray(X_binned_val)
+            X_binned_train = np.asfortranarray(X_binned_train)
+        else:
+            X_binned_train, y_train = X_binned, y
+            X_binned_val, y_val = None, None
+
+        # Subsample the training set for early stopping and score monitoring
+        if self.do_early_stopping_:
+            subsample_size = 10000  # should we expose this parameter?
+            indices = np.arange(X_binned_train.shape[0])
+            if X_binned_train.shape[0] > subsample_size:
+                # TODO: not critical but stratify using resample(stratify=y)
+                indices = rng.choice(indices, subsample_size, replace=False)
+            X_binned_small_train = X_binned_train[indices]
+            y_small_train = y_train[indices]
+            # Predicting is faster on C-contiguous arrays.
+            X_binned_small_train = np.ascontiguousarray(X_binned_small_train)
+
+        if self.verbose:
+            print("Fitting gradient boosted rounds:")
+
+        # initialize raw_predictions: those are the accumulated values
+        # predicted by the trees for the training data. raw_predictions has
+        # shape (n_trees_per_iteration, n_samples) where
+        # n_trees_per_iterations is n_classes in multiclass classification,
+        # else 1.
+        n_samples = X_binned_train.shape[0]
+        self._baseline_prediction = self.loss_.get_baseline_prediction(
+            y_train, self._n_trees_per_iteration
+        )
+        raw_predictions = np.zeros(
+            shape=(self._n_trees_per_iteration, n_samples),
+            dtype=self._baseline_prediction.dtype
+        )
+        raw_predictions += self._baseline_prediction
+
+        # initialize gradients and hessians (empty arrays).
+        # shape = (n_trees_per_iteration, n_samples).
+        gradients, hessians = self.loss_.init_gradients_and_hessians(
+            n_samples=n_samples,
+            prediction_dim=self._n_trees_per_iteration
+        )
+
+        # predictors is a matrix (list of lists) of TreePredictor objects
+        # with shape (n_iter_, n_trees_per_iteration)
+        self._predictors = predictors = []
```

Comment:
Shouldn't we make that a public fitted attribute?

In practice users might want to prune the last predictors of a fitted GBRT based on `validation_score_` so as to be able adjust the prediction speed vs accuracy trade off.

Reply:
BTW, `train_score_`  and `validation_score_` should probably be renamed to `train_scores_` and `validation_scores_`.

Reply:
I agree about the names but unfortunately the current GBDT use `train_score_` and `validation_score_`

Reply:
I'm ok with making it public but that means users have access to the `TreePredictor` class, which is supposed to be private. WDYT?

Reply:
I am not sure. Maybe we could store the record arrays with `PREDICTOR_RECORD_DTYPE` as a public attribute  and wrap those with `TreePredictor` only lazily and internally in `_raw_predict`.

At the same time it would be great to have a standard way to introspect the structure of the learned trees in scikit-learn.

Maybe it's too early to design this and we can keep this attribute private for now, although it's weird to have private fitted attribute in sklearn. It seems like an exception to me.